File: packages/ksec-shield-py/ksec_shield/integrations/langchain.py
"""
LangChain BaseCallbackHandler Integration for Agent-eBPF.
"""

from typing import Any, Dict, Optional
from ..shield import KsecShield
from ..types import KsecSecurityViolationError


class KsecLangChainCallback:
    """Callback Handler for LangChain agent runs."""
    def __init__(self, shield: KsecShield):
        self.shield = shield

    def on_tool_start(
        self,
        serialized: Dict[str, Any],
        input_str: str,
        **kwargs: Any
    ) -> None:
        tool_name = serialized.get("name", "unknown_tool")
        # Run zero-latency policy guard pass
        self.shield.guard_action(
            lambda: True,
            action_type="tool_execution",
            target=tool_name,
            metadata={"input_preview": input_str[:120]}
        )

    def on_tool_error(self, error: BaseException, **kwargs: Any) -> None:
        if isinstance(error, KsecSecurityViolationError):
            self.shield._emit_event("threat_blocked", {
                "tool_name": error.target,
                "error": str(error)
            })

File: packages/ksec-shield-ts/src/interceptors/langchain.ts
import { KsecShield } from '../index.js';

export interface LangChainToolCall {
  name: string;
  description?: string;
  [key: string]: unknown;
}

/**
 * LangChain Tracer / Callback Handler for Agent-eBPF protection.
 * Automatically wraps tool starts, LLM calls, and external requests.
 */
export class KsecLangChainCallback {
  private shield: KsecShield;

  constructor(shield: KsecShield) {
    this.shield = shield;
  }

  public async handleToolStart(tool: LangChainToolCall, input: string): Promise<void> {
    const target = tool.name;
    await this.shield.guard(
      async () => {
        // Validation pass
        return true;
      },
      {
        actionType: 'tool_execution',
        target,
        metadata: { input },
      }
    );
  }

  public async handleToolError(err: Error, tool: LangChainToolCall): Promise<void> {
    this.shield.emitCustomEvent('threat_blocked', {
      toolName: tool.name,
      error: err.message,
    });
  }
}
